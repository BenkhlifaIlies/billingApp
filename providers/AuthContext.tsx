import React from "react";
import { useState, createContext, useContext } from "react";
const useAuth = (initial: string|null) => useState<string|null>(initial);

const AuthContext = createContext<ReturnType<typeof useAuth> | null>(
  null,
);

export const useAuthContext = () => useContext(AuthContext)!;

export function AuthProvider({ children }: { children: React.ReactNode }) {
  return (
    <AuthContext.Provider value={useAuth(null)}>
      {children}
    </AuthContext.Provider>
  );
}
