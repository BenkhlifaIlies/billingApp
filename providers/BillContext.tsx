import React from "react";
import { useState, createContext, useContext } from "react";
const useBill = (initial: number) => useState<number>(initial);

const BillContext = createContext<ReturnType<typeof useBill> | null>(
  null,
);

export const useBillContext = () => useContext(BillContext)!;

export function BillProvider({ children }: { children: React.ReactNode }) {
  return (
    <BillContext.Provider value={useBill(0.05)}>
      {children}
    </BillContext.Provider>
  );
}
