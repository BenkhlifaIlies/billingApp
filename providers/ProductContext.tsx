import React from "react";
import { useState, createContext, useContext } from "react";
import { Product } from "../constants/Product";

const useProducts = (initial: Product[]) => useState<Product[]>(initial);

const ProductContext = createContext<ReturnType<typeof useProducts> | null>(
  null,
);

export const useProductContext = () => useContext(ProductContext)!;

export function ProductProvider({ children }: { children: React.ReactNode }) {
  return (
    <ProductContext.Provider value={useProducts([])}>
      {children}
    </ProductContext.Provider>
  );
}
