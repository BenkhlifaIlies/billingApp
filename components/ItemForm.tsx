import React from "react";
import {
  ActivityIndicator,
  StyleSheet,
  TextInput,
  View,
  Text,
} from "react-native";
import { Formik } from "formik";
import { Picker } from "@react-native-picker/picker";
import * as yup from "yup";
import { appleBlue, lightGrey } from "../constants/DesignColors";
import { useProductContext } from "../providers/ProductContext";
import { useBillContext } from "../providers/BillContext";
import { MAXQUANTITY } from "../constants/Calculattions";

const validationSchema = yup.object().shape({
  productLabel: yup
    .string()
    .required()
    .label("Product")
    .min(4, "Product's name is required to be at least 4 characters long"),
  productPrice: yup
    .number()
    .required()
    .label("Price")
    .test(
      "Is positive?",
      "The Price must be greater than 0",
      (value: any) => value > 0,
    )
    .typeError("The Price must be a positive number"),
  quantity: yup
    .string()
    .required()
    .label("Quantity")
    .test(
      "Is positive?",
      "Quantity must be greater than or equal to  1 and inferieur to" +
        MAXQUANTITY,
      (value: any) => value >= 1 && value < MAXQUANTITY,
    ),
});

const ItemForm = () => {
  const [productList, setProductList] = useProductContext();
  const [tax, setTax] = useBillContext();

  const handleInsertion = (
    productLabel: string,
    productPrice: number,
    quantity: number,
  ) => {
    const newArray = productList;
    try {
      newArray.unshift({
        productLabel: productLabel,
        productPrice: productPrice,
        quantity: quantity,
      });
      setProductList([...newArray]);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <View style={styles.Screenontainer}>
       <Formik
        initialValues={{ productLabel: "", productPrice: "", quantity: "" }}
        onSubmit={(values, actions) => {
          handleInsertion(
            values.productLabel,
            Number(values.productPrice),
            Number(values.quantity),
          );
          actions.resetForm();
          setTimeout(() => {
            actions.setSubmitting(false);
          }, 200);
        }}
        validationSchema={validationSchema}>
        {(formikProps) => (
          <React.Fragment>
            <View style={styles.formHeader}>
              <Text style={styles.title}>Add Product</Text>
              <Picker
                // testID="tax-picker"
                style={{
                  backgroundColor: String(appleBlue),
                  width: 250,
                  fontSize: 15,
                  fontWeight: "bold",
                  color: String(appleBlue),
                  margin: 0,
                  padding: 0,
                }}
                selectedValue={tax}
                onValueChange={(itemValue) => {
                  setTax(itemValue);
                }}>
                {/* hard coded taxes */}
                <Picker.Item label='State Selected: AK - 5.0%' value='0.05' />
                <Picker.Item label='State Selected: AL - 6.0%' value='0.06' />
                <Picker.Item label='State Selected: AR - 3.0%' value='0.03' />
                <Picker.Item label='State Selected: AS - 5.7%' value='0.057' />
                <Picker.Item label='State Selected: AZ - 10.0%' value='0.1' />
              </Picker>
            </View>
            <TextInput
              style={styles.TextInput}
              placeholder='Enter Product Label'
              value={formikProps.values.productLabel}
              onChangeText={formikProps.handleChange("productLabel")}
            />
            <Text style={styles.errorMsg}>
              {formikProps.touched.productLabel &&
                formikProps.errors.productLabel}
            </Text>
            <View
              style={[styles.container, { justifyContent: "space-between" }]}>
              <View
                style={[styles.container, { justifyContent: "flex-start" }]}>
                <View style={[styles.formHeader, { padding: 0 }]}>
                  <TextInput
                    keyboardType='numeric'
                    style={[
                      styles.TextInput,
                      styles.smallTextInput,
                      ,
                      { width: 130 },
                    ]}
                    placeholder='Enter Price'
                    value={formikProps.values.productPrice.toString()}
                    onChangeText={formikProps.handleChange("productPrice")}
                  />
                  <TextInput
                    keyboardType='numeric'
                    style={[
                      styles.TextInput,
                      styles.smallTextInput,
                      { width: 140 },
                    ]}
                    placeholder='Enter Quantity'
                    value={formikProps.values.quantity.toString()}
                    onChangeText={formikProps.handleChange("quantity")}
                  />
                </View>
              </View>
              <View
                style={{
                  display: "flex",
                  flexDirection: "column",
                  justifyContent: "center",
                }}>
                {formikProps.isSubmitting ? (
                  <ActivityIndicator color={appleBlue} />
                ) : (
                  <>
                    <Text
                      onPress={() => {
                        formikProps.handleSubmit();
                      }}
                      style={styles.clickableText}>
                      Add to list
                    </Text>
                  </>
                )}
              </View>
            </View>
            <View
              style={
                !!!formikProps.errors
                  ? { display: "none" }
                  : { display: "flex" }
              }>
              <Text style={styles.errorMsg}>
                {formikProps.touched.productPrice &&
                  formikProps.errors.productPrice}
              </Text>
              <Text style={styles.errorMsg}>
                {formikProps.touched.quantity && formikProps.errors.quantity}
              </Text>
            </View>
          </React.Fragment>
        )}
      </Formik>
    </View> 
  );
};

const styles = StyleSheet.create({
  formHeader: {
    padding: 6,
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignContent: "center",
  },
  title: {
    fontSize: 18,
    // fontWeight: "bold",
  },
  clickableText: {
    fontSize: 15,
    fontWeight: "bold",
    color: String(appleBlue),
    alignItems: "center",
    marginRight: 10,
  },
  TextInput: {
    margin: 2,
    marginVertical: 8,
    borderWidth: 1,
    borderColor: String(lightGrey),
    borderRadius: 4,
    padding: 8,
    marginBottom: 3,
    fontSize: 16,
  },
  smallTextInput: {
    width: 120,
    marginRight: 10,
  },
  errorMsg: {
    color: "red",
    fontSize: 12,
  },
  container: {
    display: "flex",
    flexDirection: "row",
  },
  Screenontainer: {
    backgroundColor: "white",
    padding: 10,
    paddingBottom: 0,
  },
});

export default ItemForm;
