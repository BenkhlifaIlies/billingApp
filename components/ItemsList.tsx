import React, { useState } from "react";
import { StyleSheet, ScrollView, Text } from "react-native";
import { DataTable } from "react-native-paper";
import { lightGrey } from "../constants/DesignColors";
import { useProductContext } from "../providers/ProductContext";
const initialEmptyPlaceHolderLength = 9;

const ItemsList = () => {
  const [productList] = useProductContext();
  const [initialEmptyPlaceHolder] = useState(
    new Array(initialEmptyPlaceHolderLength),
  );

  const renderListItems = () => {
    let i = 1;
    const t = Array.apply(
      null,
      initialEmptyPlaceHolder.slice(productList.length),
    ).map(function () {
      return (
        <DataTable.Row style={styles.tableRow} key={i++}>
          <></>
        </DataTable.Row>
      );
    });
    const result = productList.map((product) => {
      return (
        <DataTable.Row style={styles.tableRow} key={i++}>
          <DataTable.Cell>
            {product.productLabel.trim().charAt(0).toUpperCase() +
              product.productLabel.trim().slice(1)}
          </DataTable.Cell>
          <DataTable.Cell numeric>{product.quantity}</DataTable.Cell>
          <DataTable.Cell numeric>{"$" + product.productPrice}</DataTable.Cell>
          <DataTable.Cell numeric>
            {"$" + product.quantity * product.productPrice}
          </DataTable.Cell>
        </DataTable.Row>
      );
    });
    return result.concat(t);
  };

  return (
    <ScrollView>
      <DataTable>
        <DataTable.Header>
          <DataTable.Title>Name</DataTable.Title>
          <DataTable.Title numeric>Nos.</DataTable.Title>
          <DataTable.Title numeric>Price</DataTable.Title>
          <DataTable.Title numeric>Total Price</DataTable.Title>
        </DataTable.Header>
        <ScrollView style={styles.container}>{renderListItems()}</ScrollView>
      </DataTable>
    </ScrollView>
  );
  
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 10,
  },
  tableRow: {
    borderColor: "white",
    borderBottomColor: String(lightGrey),
    borderWidth: 2,
  },
});

export default ItemsList;
