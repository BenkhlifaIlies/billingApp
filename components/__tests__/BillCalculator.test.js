import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import renderer from "react-test-renderer";

import {
  BillCalculator,
  calculateTotalPriceWithoutTax,
  calculateDiscount,
  calculateTax,
  calculatetotalPriceWithTax,
} from "../BillCalculator";

import { ItemsList } from "../ItemsList";
import ItemForm from "../ItemForm";

import { BillProvider } from "../../providers/BillContext";
import { ProductProvider } from "../../providers/ProductContext";

test("matches snapshot", () => {
  const tree = renderer
    .create(
      <ProductProvider>
        <BillProvider>
          <BillCalculator />
        </BillProvider>
      </ProductProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// test("renders correctly", async() => {
//   const { getByText, getByPlaceholderText, findByText ,debug} = render(
//     <ProductProvider>
//       <BillProvider>
//         <ItemForm />
//         <ItemsList />
//         <BillCalculator />
//       </BillProvider>
//     </ProductProvider>,
//   );
//   expect(getByText(/Total price without tax/i)).not.toBeNull();
//   console.log(getByText(/Total price without tax/i));
//   await waitFor(() => {
//     fireEvent.changeText(getByPlaceholderText("Enter Product Label"),"product");
//   });
//   // debug();
//   // expect(findByText(/Total price without tax/i)).toHaveProperty("style.display","none");

//   // fireEvent.press(getByPlaceholderText("Enter Product Label"));
//   // await waitForElementToBeRemoved(() => findByText(/Total price without tax/i));

// });

describe("calculate correctly", () => {
  test("calculate total price without tax correctly", () => {
    expect(
      calculateTotalPriceWithoutTax([
        {
          productLabel: "test",
          productPrice: 100,
          quantity: 2,
        },
      ]),
    ).toBe(200);
  });
  test("calculate discount correctly", () => {
    expect(calculateDiscount(100, 0.05)).toBe(5);
  });
  test("calculate tax correctly", () => {
    expect(calculateTax(100, 0.05)).toBe(5);
  });
  test("calculate total price with taxes correctly", () => {
    expect(calculatetotalPriceWithTax(100, 0.1, 0.05)).toBe(105);
  });
});
