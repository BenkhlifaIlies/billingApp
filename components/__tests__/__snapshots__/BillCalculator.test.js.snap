// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Array [
        Object {
          "backgroundColor": "white",
          "bottom": 0,
          "padding": 10,
        },
        Object {
          "display": "flex",
        },
      ],
    ]
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginVertical": 10,
      }
    }
  >
    Total price without tax: $
    0
  </Text>
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginVertical": 10,
      }
    }
  >
    Discount 3.0%: $0.00
  </Text>
  <Text
    style={
      Object {
        "fontSize": 16,
        "marginVertical": 10,
      }
    }
  >
    Tax 5.0%: $0.00
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
          "marginVertical": 10,
        },
        Object {
          "fontWeight": "bold",
        },
      ]
    }
  >
    Total price: $0.00
  </Text>
</View>
`;
